### ISSUE TITLE
[Bug] New mark positions not applied when adding / removing lines

### ISSUE THREAD

**Issue by pidgeon777** [LeonHeidelbach/trailblazer.nvim#25]
### Did you check docs and existing issues?

- [X] I have read all the trailblazer.nvim docs
- [X] I have searched the existing issues of trailblazer.nvim
- [X] I have searched the exsiting issues of plugins related to this issue

### Neovim version (nvim -v)

v0.8.3

### Operating system/version

Windows 10 21H2

### Describe the bug

When adding or removing lines, the position of the marks under those lines is not updated accordingly, only their highlights.

When navigating to the next/previous mark, the mark highlights subsequently get updated to their wrong location, too.

It seems as if when removing/adding lines, the new mark position is computed correctly, but not really applied to the mark list, leading to a revert back to the previous mark locations.

This behaviour generates an error if for example, a mark is at the bottom of the buffer, and one line is removed. Then mark navigation produces an error because the mark is placed in a non-existent region.

### Steps To Reproduce

1. Open a `txt` file.
2. Add 2 marks on different lines.
3. Navigate between the two marks.
4. Add/remove lines between two marks.

-> New mark highlights are correctly applied.

6. Navigate between the two marks.

-> Mark highlights go back to their previous positions.

Recording of the issue:

![IGBk7BLl6Q](https://user-images.githubusercontent.com/13739892/220872557-10c07498-30ab-4b0c-8f27-6e6c477d499a.gif)

### Expected Behavior

The mark positions below the added/removed lines should be updated accordingly.

If `x` lines are added/removed, then all the marks should be shifted below/up `x` times.

It is important to note that in the previous `trailblazer.nvim` commits, this issue didn't exist.

**Comment by pidgeon777** [comment]
The issue is also there when running the `TrailBlazerMoveToNearest` command.

**Comment by pidgeon777** [comment]
The issue occurs starting from this commit:

https://github.com/LeonHeidelbach/trailblazer.nvim/commit/cda5159e430f02ec34790f3acd7ef0d94bdd09bf

**Comment by pidgeon777** [comment]
![image](https://user-images.githubusercontent.com/13739892/220886008-eb73d9c5-e5a0-4b3a-86b1-e5d9b6566f62.png)

**Comment by pidgeon777** [comment]
I switched back to the latest plugin commit, and if I change:

```lua
if ext_mark ~= nil and ext_mark[3] > 0 then
```

to

```lua
if ext_mark ~= nil and buf_list[trail_mark.buf] > ext_mark[2] + 1 then
```

then the issue seems to be solved. I'm not sure if my modification is good, though.

**Comment by LeonHeidelbach** [comment]
@pidgeon777 This was a compromise I had to make in order to enable the removal of trail marks when a line is deleted. When a line is deleted all of Neovims extmarks on that line are automatically moved to the beginning of the line. So what I do is I check if the position of a mark has changed to column 0 upon updating the trail mark stack and then remove them. The side effect is that marks that are already on column 0 will remain at their position even upon deleting the line. This is necessary to even allow placing trail marks on empty lines. I will check have a look at your solution later and see if this issue can somehow be resolved.

**Comment by pidgeon777** [comment]
I didn't know about that compromise at first.

The problem with that implementation, in addition to the fact that the marks placed in column `0` are not updated, is the following: if I create a mark in column `0` let's say of the fifth last row, and afterwards I delete more than 5 rows above the mark, then the mark remains precisely in the same line as before, which unfortunately, however, will no longer exist. This will generate an error from `Trailblazer`, as soon as I execute a command to e.g. navigate to the previous or next mark, precisely because the presence of a mark in a line that no longer exists is detected.

![image](https://user-images.githubusercontent.com/13739892/220915985-012f6ca5-26b9-4aa3-8600-1a4a9fb2dbcc.png)

Hopefully the problem of the mark placed at column `0` being incorrectly updated, and consequently the error just described when deleting rows, can be resolved somehow.

Maybe in the meanwhile it could be a good idea to add a `Known limitations` section to the readme, describing such a compromise and any other limitations?

**Comment by LeonHeidelbach** [comment]
@pidgeon777 I just fixed the error from your screenshot. I will push this change later. I am still trying to think of a solution for the 0 column issue.

**Comment by LeonHeidelbach** [comment]
I think I found a good solution for this. However, I am still working on an elegant way to enable the `move_to_nearest` functionality depending on the `fpath_up/down` directives. This requires a more involved solution than I would've anticipated.

**Comment by pidgeon777** [comment]
I just tested my modification proposed above, this is the actual behaviour.

Deleting a row containing one or two marks:

![m5kVnbmxLH](https://user-images.githubusercontent.com/13739892/220925325-e27c7648-8522-4533-838d-e8938471eb33.gif)

-> Works OK

Same thing as above, but with a mark placed at the first column of the last buffer line:

![gjRXfMR7tm](https://user-images.githubusercontent.com/13739892/220926764-2b68592d-62f7-4319-9469-57fc821c2a02.gif)

-> BAD: Mark navigation command produces an error.

Placing marks at the last line of the buffer:

![NKcvcs1wdH](https://user-images.githubusercontent.com/13739892/220927158-3c40be48-c918-4461-9ac7-aa19887f130e.gif)

-> BAD: each mark placed starting from the second column inclusive of the last buffer line, is not saved.

Summary of my modification:

- If in the last row of the buffer, a mark is placed from the 2nd column inclusive, it is displayed once it is inserted, but mark navigation leads to its deletion as if it were not applied.

- If there is a mark in the first column of the last row of the buffer, and if one or more upper rows are deleted, then mark navigation leads to the generation of an error.

In all other situations, even in the case of deletion of the last character of a row on which there may be a mark, it seems to work ok (in this case the mark is simply updated by moving it to the new last character of that row).

**Comment by LeonHeidelbach** [comment]
@pidgeon777 This should be fixed now 3f797e97272ea6b8c53712aeb2e5ea041fc1c5f7.

**Comment by pidgeon777** [comment]
Yes, I confirm it seems to work great now. Good job! 👍
